app/scripts/controllers/network.js

```diff
@@ -0,0 +1,152 @@
+  }
+
+  initializeProvider (opts) {
+    this.providerConfig = opts
+    this.provider = MetaMaskProvider(opts)
+    this.ethQuery = new EthQuery(this.provider)
+    this.lookupNetwork()
+    return Promise.resolve(this.provider)
+  }
+  switchNetwork (providerConfig) {
+    delete this.provider
+    delete this.ethQuery
+    const newConfig = extend(this.providerConfig, providerConfig)
+    this.providerConfig = newConfig
+    this.provider = MetaMaskProvider(newConfig)
+    this.ethQuery = new EthQuery(this.provider)
+    this.emit('networkSwitch', {
+      provider: this.provider,
+      ethQuery: this.ethQuery,
+    }, this.claim.bind(this))
+  }
+
+  subscribe (cb) {
+    this.networkStore.subscribe(cb)
+    this.providerStore.subscribe(cb)
+  }
+
+  verifyNetwork () {
+  // Check network when restoring connectivity:
+    if (this.isNetworkLoading()) this.lookupNetwork()
+  }
+
+  getNetworkState () {
+    return this.networkStore.getState().network
+  }
+
+  setNetworkState (network) {
+    return this.networkStore.updateState({ network })
+  }
+
+  isNetworkLoading () {
+    return this.getNetworkState() === 'loading'
+  }
+
+  lookupNetwork (err) {
+    if (err) this.setNetworkState('loading')
+
+    this.ethQuery.sendAsync({ method: 'net_version' }, (err, network) => {
+      if (err) return this.setNetworkState('loading')
+
+      log.info('web3.getNetwork returned ' + network)
+      this.setNetworkState(network)
+    })
+  }
+
+  setRpcTarget (rpcUrl) {
+    this.providerStore.updateState({
+      provider: {
+        type: 'rpc',
+        rpcTarget: rpcUrl,
+      },
+    })
+  }
+
+  getCurrentRpcAddress () {
+    var provider = this.getProvider()
+    if (!provider) return null
+    return this.getRpcAddressForType(provider.type)
+  }
+
+  setProviderType (type) {
+    if (type === this.getProvider().type) return
+    const rpcTarget = this.getRpcAddressForType(type)
+    this.networkStore.updateState({network: 'loading'})
+    this.switchNetwork({
+      rpcUrl: rpcTarget,
+    })
+    this.once('claimed', () => {
+      this.providerStore.updateState({provider: {type, rpcTarget}})
+      console.log('CLAIMED')
+      this.lookupNetwork()
+    })
+
+  }
+
+  useEtherscanProvider () {
+    this.setProviderType('etherscan')
+  }
+
+  getProvider () {
+    return this.providerStore.getState().provider
+  }
+
+  getRpcAddressForType (type) {
+    const provider = this.getProvider()
+    switch (type) {
+
+      case 'mainnet':
+        return MAINNET_RPC
+
+      case 'testnet':
+        return TESTNET_RPC
+
+      case 'morden':
+        return MORDEN_RPC
+
+      case 'kovan':
+        return KOVAN_RPC
+
+      case 'rinkeby':
+        return RINKEBY_RPC
+
+      default:
+        return provider && provider.rpcTarget ? provider.rpcTarget : TESTNET_RPC
+    }
+  }
+
+  claim () {
+    this._claimed += 1
+    if (this._claimed === this.listenerCount('networkSwitch')) {
+      this.emit('claimed')
+      this._claimed = 0
+    }
```

Comment:
since ethStore only updates the account when it has seen a block [here]( https://github.com/MetaMask/metamask-plugin/blob/master/app/scripts/lib/eth-store.js#L42
switching networks)  I wanted to know when when every thing had "claimed" the new provider so i could change the network id from `loading` to have a better effect on the UI. This way the loading spinner would stop only when ethStore had seen a block and was looking up balances. otherwise you switch networks and you balance would be `...` and then it would load a balance. But i will do this a different way then